[AGNTLOG-56 ] Create Auditor Component

soberpeach opened the pull request:
<!--
* Contributors are encouraged to read our [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* Both Contributor and Reviewer Checklists are available at https://datadoghq.dev/datadog-agent/guidelines/contributing/#pull-requests.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
* Please fill the below sections if possible with relevant information or links.
-->
### What does this PR do?

This is the first a series of PRs that will migrate the auditor to a component. This is the first PR and focuses only on creating the auditor component.

Subsequent PRs will handle:
1. Integrating the new auditor component into the codebase, this would involve replacing instances of the existing auditor in `pkg/logs/auditor` with the new component this PR creates
    * This could be done incrementally, starting with the base agent and then expanding to serverless, etc.
2. Adding functionality for the auditor to mitigate memory leaks in Fargate.
 3. Removing the old auditor from the codebase.

### Motivation

This change is a part of addressing a memory leak in ECS Fargate, as well as a a part of a more general push to move more parts of the agent to components.

### Describe how you validated your changes
<!--
Validate your changes before merge, ensuring that:
* Your PR is tested by static / unit / integrations / e2e tests
* Your PR description details which e2e tests cover your changes, if any
* The PR description contains details of how you validated your changes. If you validated changes manually and not through automated tests, add context on why automated tests did not fit your changes validation.

If you want additional validation by a second person, you can ask reviewers to do it. Describe how to set up an environment for manual tests in the PR description. Manual validation is expected to happen on every commit before merge.

Any manual validation step should then map to an automated test. Manual validation should not substitute automation, minus exceptions not supported by test tooling yet.
-->

This change purely adds the auditor component, and doesn't do anything with it, later changes will deal with replacing the old auditor in `pkg/logs` and ensuring functionality is consistent.

The unit tests from the old auditor have also been ported over and ensure identical functionality of the component.

### Possible Drawbacks / Trade-offs

### Additional Notes
<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

soberpeach committed:
Create auditor component

gh123man commented:
IMO if we are introducing new code, we should introduce new tests with it. 
Is it possible to include the related UTs in this PR?

I assume you have a plan for this (tests, and replacing the existing code), could you add more details about your plan for future PRs to the description?

soberpeach committed:
Fixed public function comment

soberpeach committed:
Added testing for auditor component

soberpeach committed:
Merge branch 'main' into lucas.liseth/create-auditor-component

soberpeach committed:
Migrate api tests to component implementation

gh123man reviewed (changes requested):
Just some minor comments

soberpeach commented on comp/logs/auditor/impl-none/auditor_noop.go:
```diff
@@ -0,0 +1,66 @@
+		case <-a.channel:
+		// drain the channel, we're not doing anything with the channel
+		case <-a.stopChannel:
+			// TODO(remy): close the message channel
```
Addressed! I just closed the channel, not sure if that was the right way to do it or I should put it in a `defer` statement, lmk.

soberpeach commented on comp/logs/auditor/impl/auditor_test.go:
```diff
@@ -0,0 +1,132 @@
+		Log:    logComponent,
+	}
+
+	suite.a = NewAuditor(deps).Comp
```
Ah I've seen that pattern before but didn't realize it was for this purpose. Changed it.

soberpeach committed:
Address PR comments

1. Closed channel in the noop auditor
2. Refactored the auditor constructor to two public and private constructors so
the private constructor can be used in tests.
3. Removed extraneous TODO from old api_v0

soberpeach commented on comp/logs/auditor/impl/api_v0.go:
```diff
@@ -0,0 +1,47 @@
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//nolint:revive // TODO(AML) Fix revive linter
```
Removed the TODO as there's already a package level comment 👍

soberpeach commented:
/merge
